mithril-common/src/chain_observer/pallas_observer.rs

```diff
@@ -239,6 +231,45 @@ impl PallasChainObserver {
         Ok(Some(stake_distribution))
     }
 
+    /// Fetches chain point and genesis config through the local statequery.
+    /// The KES period is calculated afterwards.
+    async fn calculate_kes_period(
+        &self,
+        client: &mut NodeClient,
+    ) -> Result<Option<KESPeriod>, ChainObserverError> {
+        let statequery = client.statequery();
+
+        statequery
+            .acquire(None)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to acquire statequery")?;
+
+        let chain_point = queries_v16::get_chain_point(statequery)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to get chain point")?;
+
+        let era = queries_v16::get_current_era(statequery)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to get current era")?;
+
+        let genesis_config = queries_v16::get_genesis_config(statequery, era)
+            .await
+            .map_err(|err| anyhow!(err))
+            .with_context(|| "PallasChainObserver failed to get genesis config")?;
+
+        let config = genesis_config
+            .first()
+            .with_context(|| "PallasChainObserver failed to extract the config")?;
+
+        let slots_per_kes_period = config.slots_per_kes_period as u64;
+        let current_kes_period = chain_point.slot_or_default() / slots_per_kes_period;
```

Comment:
If `slots_per_kes_period` is `0`, then it will create a panic, can you send an error instead?

Reply:
done